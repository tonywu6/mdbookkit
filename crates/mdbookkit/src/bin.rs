#[cfg(feature = "lib-rustdoc-link")]
pub mod rustdoc_link;
