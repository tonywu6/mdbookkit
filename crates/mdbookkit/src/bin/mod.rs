#[cfg(feature = "lib-rustdoc-link")]
pub mod rustdoc_link;

#[cfg(feature = "lib-link-forever")]
pub mod link_forever;
