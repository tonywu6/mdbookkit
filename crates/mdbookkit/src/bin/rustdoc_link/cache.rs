use std::{
    borrow::Cow,
    collections::{BTreeSet, HashMap},
    hash::Hash,
    iter,
};

use anyhow::{bail, Context, Result};
use lsp_types::Url;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use sha2::{Digest, Sha256};
use tap::{Pipe, Tap, TapFallible};
use tokio::task::JoinSet;

use crate::log_debug;

use super::{env::Environment, link::ItemLinks, page::Pages, url::UrlToPath, Resolver};

#[allow(async_fn_in_trait)]
pub trait Cache: DeserializeOwned + Serialize {
    type Validated: Resolver;

    async fn reuse(self, env: &Environment) -> Result<Self::Validated>;

    async fn build<K>(env: &Environment, content: &Pages<'_, K>) -> Result<Self>
    where
        K: Eq + Hash;

    async fn load(env: &Environment) -> Result<Self::Validated> {
        env.load_temp::<Self, _>("cache.json")
            .tap_err(log_debug!())?
            .reuse(env)
            .await
            .tap_err(log_debug!())
    }

    async fn save<K>(env: &Environment, content: &Pages<'_, K>) -> Result<()>
    where
        K: Eq + Hash,
    {
        let this = Self::build(env, content).await?;
        env.save_temp::<Self, _>("cache.json", &this)
            .tap_err(log_debug!())
    }
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum FileCache {
    V1(FileCacheV1),
}

impl Cache for FileCache {
    type Validated = HashMap<String, ItemLinks>;

    async fn reuse(self, env: &Environment) -> Result<Self::Validated> {
        match self {
            Self::V1(cache) => Ok(cache.reuse(env).await?),
        }
    }

    async fn build<K>(env: &Environment, content: &Pages<'_, K>) -> Result<Self>
    where
        K: Eq + Hash,
    {
        Ok(Self::V1(FileCacheV1::build(env, content).await?))
    }
}

#[derive(Debug, Deserialize, Serialize)]
pub struct FileCacheV1 {
    hash: String,
    urls: Vec<(String, ItemLinks)>,
}

impl Cache for FileCacheV1 {
    type Validated = HashMap<String, ItemLinks>;

    async fn reuse(self, env: &Environment) -> Result<Self::Validated> {
        let deps = self
            .urls
            .iter()
            .flat_map(|(_, links)| links.deps())
            .map(Cow::Borrowed);

        let hash = Self::hash(env, deps).await;

        if hash != self.hash {
            bail!("checksum mismatch, expected {}, actual {hash}", self.hash)
        }

        Ok(self.urls.into_iter().collect())
    }

    async fn build<K>(env: &Environment, content: &Pages<'_, K>) -> Result<Self>
    where
        K: Eq + Hash,
    {
        let urls = content
            .links()
            .iter()
            .map(|(k, v)| (k.to_string(), v.clone()))
            .collect::<Vec<_>>();

        let deps = urls
            .iter()
            .flat_map(|(_, links)| links.deps())
            .map(Cow::Borrowed);

        let hash = Self::hash(env, deps).await;

        Ok(Self { hash, urls })
    }
}

impl FileCacheV1 {
    async fn hash<'a, D>(env: &'a Environment, deps: D) -> String
    where
        D: Iterator<Item = Cow<'a, Url>>,
    {
        iter::once(Cow::Owned(env.source_dir.join("Cargo.toml").unwrap()))
            .chain(iter::once(Cow::Owned(
                env.crate_dir.join("Cargo.toml").unwrap(),
            )))
            .chain(iter::once(Cow::Borrowed(&env.entrypoint)))
            .chain(deps)
            .collect::<BTreeSet<_>>()
            .into_iter()
            .filter_map(|dep| {
                if env.source_dir.make_relative(&dep)?.starts_with("../") {
                    None
                } else {
                    Some(dep.into_owned())
                }
            })
            .map(read_dep)
            .collect::<JoinSet<_>>()
            .join_all()
            .await
            .into_iter()
            .filter_map(|result| {
                result
                    .context("failed to read cache dependency")
                    .tap_err(log_debug!())
                    .ok()
            })
            .collect::<Vec<_>>()
            .tap_mut(|deps| deps.sort_by(|(k1, _), (k2, _)| k1.cmp(k2)))
            .into_iter()
            .fold(Sha256::new(), |mut hash, (_, src)| {
                hash.update(src);
                hash
            })
            .pipe(|hash| hash.finalize().digest())
    }
}

async fn read_dep(url: Url) -> Result<(String, String)> {
    let content = tokio::fs::read_to_string(&url.to_path()?).await?;
    Ok((url.to_string(), content))
}

trait HexDigest {
    fn digest(&self) -> String;
}

impl HexDigest for sha2::digest::Output<Sha256> {
    fn digest(&self) -> String {
        format!("{:064x}", self)
    }
}
