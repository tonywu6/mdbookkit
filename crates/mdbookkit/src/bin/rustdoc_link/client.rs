use std::{
    collections::HashMap,
    ops::ControlFlow,
    process::Stdio,
    sync::{Arc, RwLock},
    task::Poll,
    time::Duration,
};

use anyhow::{anyhow, bail, Context, Result};
use async_lsp::{
    concurrency::ConcurrencyLayer, panic::CatchUnwindLayer, router::Router, LanguageServer,
    MainLoop, ServerSocket,
};
use lsp_types::{
    notification::{LogMessage, Progress, PublishDiagnostics, ShowMessage},
    request::Request,
    ClientCapabilities, ClientInfo, DidCloseTextDocumentParams, DidOpenTextDocumentParams,
    GeneralClientCapabilities, GotoDefinitionParams, GotoDefinitionResponse, InitializeParams,
    InitializeResult, InitializedParams, LogMessageParams, MessageType, NumberOrString, Position,
    PositionEncodingKind, ProgressParams, ProgressParamsValue, ServerInfo, ShowMessageParams,
    TextDocumentIdentifier, TextDocumentItem, TextDocumentPositionParams, Url,
    WindowClientCapabilities, WorkDoneProgress, WorkspaceFolder,
};
use serde::{Deserialize, Serialize};
use serde_json::json;
use tap::{Pipe, TapFallible};
use tokio::{
    sync::{mpsc, OnceCell, OwnedSemaphorePermit, Semaphore},
    task::JoinHandle,
};
use tokio_util::compat::{TokioAsyncReadCompatExt, TokioAsyncWriteCompatExt};
use tower::ServiceBuilder;

use crate::{log_debug, log_warning, logging::spinner};

use super::{
    env::Environment,
    link::ItemLinks,
    sync::{EventSampler, EventSampling},
    url::UrlToPath,
};

/// LSP client to talk to rust-analyzer.
///
/// [`Client`] does not implement [`Clone`], because the server instance is lazily spawned
/// and so must be unique for each client. To enable cloning, put this in an [`Arc`].
#[derive(Debug)]
pub struct Client {
    env: Environment,
    server: OnceCell<Server>,
    docs: DocumentLock,
}

impl Client {
    /// Initialize a new LSP client.
    ///
    /// This does not actually spawn rust-analyzer.
    pub fn new(env: Environment) -> Self {
        let server = OnceCell::new();
        let docs = DocumentLock::default();
        Self { env, server, docs }
    }

    pub fn env(&self) -> &Environment {
        &self.env
    }

    pub async fn open(&self, uri: Url, text: String) -> Result<OpenDocument> {
        let server = self
            .server
            .get_or_try_init(|| Server::spawn(&self.env))
            .await?
            .clone();

        let opened = self.docs.open(server.server.clone(), uri, text).await?;

        server
            .stabilizer
            .wait()
            .await
            .with_context(|| format!("using rust-analyzer version {}", ra_version(&server.info)))
            .context("timed out waiting for rust-analyzer to finish indexing")?;

        Ok(opened)
    }

    /// Shutdown the server if it was spawned.
    ///
    /// Returns the [`Environment`] struct for further use.
    ///
    /// This moves `self`. To shutdown a [`Client`] that's in an [`Arc`],
    /// use [`client.drop`][Client::drop].
    pub async fn stop(self) -> Environment {
        if let Some(server) = self.server.into_inner() {
            server
                .dispose()
                .await
                .context("failed to properly stop rust-analyzer")
                .tap_err(log_warning!())
                .ok();
        }
        self.env
    }

    /// See [`client.stop`][Client::stop].
    pub async fn drop(self: Arc<Self>) -> Result<Environment> {
        let Some(this) = Arc::into_inner(self) else {
            bail!("attempted to shutdown a client that is still referenced")
        };
        Ok(this.stop().await)
    }
}

#[derive(Debug, Clone)]
struct Server {
    server: ServerSocket,
    stabilizer: EventSampler<()>,
    background: Arc<JoinHandle<()>>,
    info: Option<ServerInfo>,
}

impl Server {
    async fn spawn(env: &Environment) -> Result<Self> {
        macro_rules! ra_spinner {
            () => {
                "rust-analyzer"
            };
        }

        /// Listen for [Work Done Progress][workDoneProgress] events from rust-analyzer
        /// to determine if indexing is done.
        ///
        /// The `cachePriming` events look like this:
        ///
        /// - [`WorkDoneProgress::Begin`]
        /// - [`WorkDoneProgress::Report`] for each crate indexed, with messages like `4/200 (core)`
        /// - ...
        /// - [`WorkDoneProgress::End`]
        ///
        /// (This is the ticker thing that shows up in the VS Code status bar).
        ///
        /// Notably, rust-analyzer seems to do several rounds of indexing, **not all of
        /// which is actually indexing the entire codebase:**
        ///
        /// - First, a `0/1 (<crate name>)` round that only indexes the current crate.
        ///
        ///   - RA is **not ready** at this point: if we were to query for external docs
        ///     immediately after this [`WorkDoneProgress::End`] event, almost all links
        ///     will fail to resolve.
        ///
        /// - Then, a `0/x ...` round that seems to actually indexes everything, including
        ///   [`std`] and all the dependencies.
        ///
        ///   - RA is likely ready at this point.
        ///
        /// - Then, one or more additional rounds of indexing that finish very quickly.
        ///
        /// To be able to "reliably" determine that RA is ready for querying, the
        /// probing mechanism does essentially the following:
        ///
        /// 1. Ignore the first round of indexing.
        /// 2. Be extra pessimistic and wait for events to quiet down after receiving
        ///    the first [`WorkDoneProgress::End`] event; see [`EventSampler`].
        ///
        /// [workDoneProgress]: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workDoneProgress
        fn probe_progress(state: &mut State, progress: ProgressParams) {
            match indexing_progress(&progress) {
                Some(WorkDoneProgress::Begin(begin)) => {
                    state.percent_indexed = Some(0);

                    let msg = begin.message.as_deref().unwrap_or_default();
                    spinner().update(ra_spinner!(), msg);

                    let tx = state.tx.clone();
                    tokio::spawn(async move { tx.send(Poll::Pending).await.ok() });
                }

                Some(WorkDoneProgress::Report(report)) => {
                    if let Some(msg) = &report.message {
                        spinner().update(ra_spinner!(), msg);
                    }

                    let Some(indexed) = state.percent_indexed.as_mut() else {
                        // progress was invalidated
                        return;
                    };

                    let update = report.percentage.unwrap_or(0);

                    let spurious = if let Some(msg) = &report.message {
                        // HACK: invalidate progress reports that say "0/1 (crate name)"
                        // because RA isn't actually indexing everything at this point
                        msg.starts_with("0/1 (")
                    } else {
                        // also ignore indexing runs that went from 0 to a 100
                        *indexed == 0 && update == 100
                    };

                    if spurious {
                        log::debug!("ignoring spurious rust-analyzer progress");
                        state.percent_indexed = None;
                        return;
                    }

                    if update >= *indexed {
                        *indexed = update;
                    }
                }

                Some(WorkDoneProgress::End(end)) => {
                    let Some(indexed) = state.percent_indexed else {
                        // progress was invalidated
                        return;
                    };

                    if indexed < 99 {
                        return;
                    }

                    let msg = end.message.as_deref().unwrap_or("indexing done");
                    spinner().update(ra_spinner!(), msg);

                    let tx = state.tx.clone();
                    tokio::spawn(async move { tx.send(Poll::Ready(())).await.ok() });
                }

                None => {
                    log::trace!("{progress:#?}")
                }
            }
        }

        let (tx, rx) = mpsc::channel(16);

        let stabilizer = EventSampling {
            buffer: Duration::from_millis(500),
            timeout: env.config.rust_analyzer_timeout(),
        }
        .using(rx);

        struct State {
            tx: mpsc::Sender<Poll<()>>,
            percent_indexed: Option<u32>,
        }

        let (background, mut server) = MainLoop::new_client(move |_| {
            let state = State {
                tx,
                percent_indexed: Some(0),
            };

            let mut router = Router::new(state);

            router
                .notification::<Progress>(|state, progress| {
                    log::trace!("{progress:#?}");
                    probe_progress(state, progress);
                    ControlFlow::Continue(())
                })
                .notification::<PublishDiagnostics>(|_, diagnostics| {
                    log::trace!("{diagnostics:#?}");
                    ControlFlow::Continue(())
                })
                .notification::<ShowMessage>(|_, ShowMessageParams { typ, message }| {
                    log_message(message, typ);
                    ControlFlow::Continue(())
                })
                .notification::<LogMessage>(|_, LogMessageParams { typ, message }| {
                    log_message(message, typ);
                    ControlFlow::Continue(())
                })
                .event::<StopEvent>(|_, _| ControlFlow::Break(Ok(())));

            ServiceBuilder::new()
                .layer(CatchUnwindLayer::default())
                .layer(ConcurrencyLayer::default())
                .service(router)
        });

        let proc = env
            .command()?
            .current_dir(env.crate_dir.to_path()?)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::inherit())
            .spawn()
            .context("failed to spawn rust-analyzer")?;

        let background = tokio::spawn(async move {
            let mut proc = proc;
            let stdout = proc.stdout.take().unwrap();
            let stdin = proc.stdin.take().unwrap();
            background
                .run_buffered(stdout.compat(), stdin.compat_write())
                .await
                .tap_err(log_debug!())
                .ok();
        })
        .pipe(Arc::new);

        let features = {
            let features = &env.config.cargo_features;
            if features.len() == 1 && features[0] == "all" {
                json!("all")
            } else {
                json!(features)
            }
        };

        let InitializeResult {
            server_info: info,
            capabilities,
        } = server
            .initialize(InitializeParams {
                workspace_folders: Some(vec![WorkspaceFolder {
                    uri: env.crate_dir.clone(),
                    name: "root".into(),
                }]),

                capabilities: ClientCapabilities {
                    experimental: Some(json! {{
                        "localDocs": true,
                    }}),
                    window: Some(WindowClientCapabilities {
                        work_done_progress: Some(true),
                        ..Default::default()
                    }),
                    general: Some(GeneralClientCapabilities {
                        position_encodings: Some(vec![PositionEncodingKind::UTF8]),
                        ..Default::default()
                    }),
                    ..Default::default()
                },

                initialization_options: Some(json! {{
                    "cachePriming": {
                        "enable": true,
                        "numThreads": "physical",
                    },
                    "cargo": {
                        "features": features,
                    }
                }}),

                client_info: Some(ClientInfo {
                    name: env!("CARGO_PKG_NAME").into(),
                    version: Some(env!("CARGO_PKG_VERSION").into()),
                }),

                ..Default::default()
            })
            .await?;

        if capabilities.position_encoding != Some(PositionEncodingKind::UTF8) {
            let version = ra_version(&info);
            let error = anyhow!("using rust-analyzer version {version}")
                .context("this rust-analyzer does not support utf-8 positions");
            bail!(error)
        }

        server.initialized(InitializedParams {})?;

        spinner().create(ra_spinner!(), None);

        Ok(Self {
            server,
            stabilizer,
            background,
            info,
        })
    }

    async fn dispose(self) -> Result<()> {
        let Self {
            mut server,
            background,
            ..
        } = self;
        let background = Arc::into_inner(background)
            .expect("should not dispose while multiple server sockets are still alive");
        server.shutdown(()).await?;
        server.exit(())?;
        server.emit(StopEvent)?;
        background.await?;
        Ok(())
    }
}

struct StopEvent;

#[derive(Debug, Default)]
struct DocumentLock {
    opened: RwLock<DocumentMap>,
}

type DocumentMap = HashMap<Url, (Arc<Semaphore>, i32)>;

impl DocumentLock {
    pub async fn open(
        &self,
        mut server: ServerSocket,
        uri: Url,
        text: String,
    ) -> Result<OpenDocument> {
        let (sema, version) = {
            let mut lock = self.opened.write().unwrap();
            let (sema, version) = lock
                .entry(uri.clone())
                .or_insert_with(|| (Arc::new(Semaphore::new(1)), 0));
            *version += 1;
            (sema.clone(), *version)
        };

        let _permit = sema.acquire_owned().await?;

        server.did_open(DidOpenTextDocumentParams {
            text_document: TextDocumentItem {
                language_id: "rust".into(),
                uri: uri.clone(),
                text,
                version,
            },
        })?;

        log::debug!("textDocument/didOpen {}", uri);

        Ok(OpenDocument {
            uri,
            server,
            _permit,
        })
    }
}

/// [OpenDocument] does not implement [Clone] because it will cause extraneous
/// `textDocument/didClose` notifications. Cloning should be done using [Arc].
#[derive(Debug)]
#[must_use = "bind this to a variable or document will be immediately closed"]
pub struct OpenDocument {
    uri: Url,
    server: ServerSocket,
    _permit: OwnedSemaphorePermit,
}

impl OpenDocument {
    pub async fn resolve(&self, position: Position) -> Result<ItemLinks> {
        let defs = self
            .server
            .clone()
            .definition(GotoDefinitionParams {
                text_document_position_params: document_position(self.uri.clone(), position),
                work_done_progress_params: Default::default(),
                partial_result_params: Default::default(),
            })
            .await
            .context("failed to request source definition")
            .tap_err(log_warning!())
            .unwrap_or_default()
            .map(|defs| match defs {
                GotoDefinitionResponse::Scalar(loc) => vec![loc.uri],
                GotoDefinitionResponse::Array(locs) => {
                    locs.into_iter().map(|loc| loc.uri).collect()
                }
                GotoDefinitionResponse::Link(links) => {
                    links.into_iter().map(|link| link.target_uri).collect()
                }
            })
            .unwrap_or_default();

        let ExternalDocLinks { web, local } = self
            .server
            .request::<ExternalDocs>(document_position(self.uri.clone(), position))
            .await
            .context("failed to request external docs")
            .tap_err(log_warning!())
            .unwrap_or_default()
            .context("server returned no result for external docs")?;

        ItemLinks::new(web, local, defs)
    }
}

impl Drop for OpenDocument {
    fn drop(&mut self) {
        self.server
            .did_close(DidCloseTextDocumentParams {
                text_document: TextDocumentIdentifier {
                    uri: self.uri.clone(),
                },
            })
            .tap_ok(|_| log::debug!("textDocument/didClose {}", self.uri))
            .context("error sending textDocument/didClose")
            .tap_err(log_debug!())
            .ok();
    }
}

enum ExternalDocs {}

impl Request for ExternalDocs {
    const METHOD: &'static str = "experimental/externalDocs";
    type Params = TextDocumentPositionParams;
    type Result = Option<ExternalDocLinks>;
}

#[derive(Debug, Deserialize, Serialize)]
struct ExternalDocLinks {
    web: Option<Url>,
    local: Option<Url>,
}

fn document_position(uri: Url, position: Position) -> TextDocumentPositionParams {
    TextDocumentPositionParams {
        text_document: TextDocumentIdentifier { uri },
        position,
    }
}

fn indexing_progress(progress: &ProgressParams) -> Option<&WorkDoneProgress> {
    match progress {
        ProgressParams {
            token: NumberOrString::String(token),
            value: ProgressParamsValue::WorkDone(progress),
        } if matches!(
            token.as_ref(),
            "rustAnalyzer/Indexing" | "rustAnalyzer/cachePriming"
        ) =>
        {
            Some(progress)
        }
        _ => None,
    }
}

fn log_message(message: String, typ: MessageType) {
    match typ {
        MessageType::ERROR | MessageType::WARNING => {
            log::warn!("rust-analyzer: {message}")
        }
        MessageType::INFO | MessageType::LOG => {
            log::debug!("rust-analyzer: {message}")
        }
        _ => log::trace!("rust-analyzer: {message}"),
    }
}

fn ra_version(info: &Option<ServerInfo>) -> &str {
    info.as_ref()
        .and_then(|ServerInfo { version, .. }| version.as_deref())
        .unwrap_or("(unknown)")
}
