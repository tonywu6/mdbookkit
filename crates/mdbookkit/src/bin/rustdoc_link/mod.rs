use std::{borrow::Borrow, collections::HashMap, hash::Hash, sync::Arc};

use anyhow::{Context, Result};
use lsp_types::Position;
use tap::{Pipe, TapFallible};
use tokio::task::JoinSet;

mod client;
pub mod env;
mod item;
mod link;
mod markdown;
mod page;
mod sync;

#[cfg(feature = "rustdoc-link")]
pub mod cache;

use crate::{log_debug, logging::spinner, styled};

pub use self::{client::Client, page::Pages};
use self::{item::Item, link::ItemLinks};

/// Type that can provide links.
///
/// Resolvers should modify the provided [`Pages`] in place.
///
/// This is currently an abstraction over two sources of links:
///
/// - [`Client`], which invokes rust-analyzer
/// - [`Cache`][crate::cache::Cache] implementations
#[allow(async_fn_in_trait)]
pub trait Resolver {
    async fn resolve<K>(&self, pages: &mut Pages<'_, K>) -> Result<()>
    where
        K: Eq + Hash;
}

impl Resolver for Client {
    async fn resolve<K>(&self, pages: &mut Pages<'_, K>) -> Result<()>
    where
        K: Eq + Hash,
    {
        let request = pages.items();

        if request.is_empty() {
            return Ok(());
        }

        let main = std::fs::read_to_string(self.env().entrypoint.path())?;

        let (context, request) = {
            let mut context = format!("{main}\nfn {UNIQUE_ID} () {{\n");

            let line = context.chars().filter(|&c| c == '\n').count();

            let request = request
                .iter()
                .scan(line, |line, (key, item)| {
                    build(&mut context, line, item).map(|cursors| (key.clone(), cursors))
                })
                .collect::<Vec<_>>();

            fn build(context: &mut String, line: &mut usize, item: &Item) -> Option<Vec<Position>> {
                use std::fmt::Write;
                let _ = writeln!(context, "{}", item.stmt);
                let cursors = item
                    .cursor
                    .as_ref()
                    .iter()
                    .map(|&col| Position::new(*line as _, col as _))
                    .collect::<Vec<_>>();
                *line += 1;
                Some(cursors)
            }

            context.push('}');

            (context, request)
        };

        log::debug!("request context\n\n{context}\n");

        let document = self
            .open(self.env().entrypoint.clone(), context)
            .await?
            .pipe(Arc::new);

        spinner().create("resolve", Some(request.len() as _));

        let tasks: JoinSet<Option<(String, ItemLinks)>> = request
            .into_iter()
            .map(|(key, pos)| {
                let key = key.to_string();
                let doc = document.clone();
                resolve(doc, key, pos)
            })
            .collect();

        async fn resolve(
            doc: Arc<client::OpenDocument>,
            key: String,
            pos: Vec<Position>,
        ) -> Option<(String, ItemLinks)> {
            let _task = spinner().task("resolve", &key);
            for p in pos {
                let resolved = doc
                    .resolve(p)
                    .await
                    .with_context(|| format!("error resolving {p:?}"))
                    .tap_err(log_debug!())
                    .ok();
                if let Some(resolved) = resolved {
                    return Some((key, resolved));
                }
            }
            None
        }

        let resolved = tasks
            .join_all()
            .await
            .into_iter()
            .flatten()
            .collect::<HashMap<_, _>>();

        spinner().finish("resolve", styled!(("done").green()));

        pages.apply(&resolved);

        Ok(())
    }
}

impl<K> Resolver for HashMap<K, ItemLinks>
where
    K: Borrow<str> + Eq + Hash,
{
    async fn resolve<P>(&self, pages: &mut Pages<'_, P>) -> Result<()>
    where
        P: Eq + Hash,
    {
        pages.apply(self);
        Ok(())
    }
}

const UNIQUE_ID: &str = "__ded48f4d_0c4f_4950_b17d_55fd3b2a0c86__";
