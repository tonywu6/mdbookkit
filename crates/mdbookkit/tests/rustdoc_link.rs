use std::{io::Write, sync::Arc};

use anyhow::{bail, Context, Result};

use assert_cmd::{prelude::*, Command};
use predicates::prelude::*;
use similar::{ChangeTag, TextDiff};
use tap::Pipe;
use tempfile::TempDir;
use tokio::task::JoinSet;

use mdbookkit::bin::rustdoc_link::{
    env::{find_code_extension, Config, Environment},
    Client, Pages, Resolver,
};
use util_testing::{may_skip, portable_snapshots, setup_paths, test_document, TestDocument};

mod util;

async fn snapshot(
    client: Arc<Client>,
    TestDocument { source, name, .. }: TestDocument,
) -> Result<()> {
    let stream = client.env().markdown(source).into_offset_iter();

    let mut page = Pages::one(source, stream)?;

    client.resolve(&mut page).await?;

    let output = page.get(&client.env().emit_config())?.to_string();

    portable_snapshots!().test(|| insta::assert_snapshot!(name.clone(), output))?;

    assert_no_whitespace_change(source, &output)?;

    let report = page
        .reporter()
        .level(log::LevelFilter::Info)
        .names(|_| name.clone())
        .colored(false)
        .logging(false)
        .build()
        .to_report();

    portable_snapshots!().test(|| insta::assert_snapshot!(format!("{name}.stderr"), report))?;

    Ok(())
}

fn assert_no_whitespace_change(source: &str, output: &str) -> Result<()> {
    let changed_lines = TextDiff::from_words(source, output)
        .iter_all_changes()
        .filter_map(|change| {
            if matches!(change.tag(), ChangeTag::Equal) {
                return None;
            }
            if change.value().contains('\n') {
                Some(change.value())
            } else {
                None
            }
        })
        .collect::<Vec<_>>();

    if !changed_lines.is_empty() {
        bail!("unexpected whitespace change: {changed_lines:?}")
    } else {
        Ok(())
    }
}

#[tokio::test]
async fn test_snapshots() -> Result<()> {
    util::setup_logging();

    let client = client()?;

    let tests = [
        test_document!("../../../docs/src/rustdoc-link/supported-syntax.md"),
        test_document!("../../../docs/src/rustdoc-link/known-issues.md"),
        test_document!("../../../docs/src/rustdoc-link/getting-started.md"),
        test_document!("../../../docs/src/rustdoc-link/index.md"),
        test_document!("tests/ra-known-quirks.md"),
    ];

    let errors = tests
        .map(|test| snapshot(client.clone(), test))
        .into_iter()
        .collect::<JoinSet<_>>()
        .join_all()
        .await
        .into_iter()
        .filter_map(Result::err)
        .collect::<Vec<_>>();

    if !errors.is_empty() {
        let errors = errors
            .iter()
            .map(|e| format!("{e:?}"))
            .collect::<Vec<_>>()
            .join("\n");
        panic!("{errors}")
    }

    client.drop().await?;

    Ok(())
}

fn client() -> Result<Arc<Client>> {
    Config {
        rust_analyzer: Some("cargo run --package util-rust-analyzer -- analyzer".into()),
        cargo_features: vec!["rustdoc-link".into()],
        ..Default::default()
    }
    .pipe(Environment::new)?
    .pipe(Client::new)
    .pipe(Arc::new)
    .pipe(Ok)
}

#[test]
#[ignore = "should run in a dedicated environment"]
fn test_minimum_env() -> Result<()> {
    util::setup_logging();

    log::info!("setup: compile self");
    Command::new("cargo")
        .args([
            "build",
            "--package",
            env!("CARGO_PKG_NAME"),
            "--all-features",
            "--bin",
            "mdbook-rustdoc-link",
        ])
        .arg(if cfg!(debug_assertions) {
            "--profile=dev"
        } else {
            "--profile=release"
        })
        .assert()
        .success();

    let path = setup_paths()?;

    let root = TempDir::new()?;

    log::debug!("{root:?}");

    log::info!("given: a book");
    Command::new("mdbook")
        .args(["init", "--force"])
        .env("PATH", &path)
        .current_dir(&root)
        .unwrap()
        .assert()
        .success();

    log::info!("given: preprocessor is enabled");
    std::fs::File::options()
        .append(true)
        .open(root.path().join("book.toml"))?
        .pipe(|mut file| file.write_all("[preprocessor.rustdoc-link]\n".as_bytes()))?;

    log::info!("when: book is not a Cargo project");
    log::info!("then: preprocessor fails");
    Command::new("mdbook")
        .arg("build")
        .env("PATH", &path)
        .current_dir(&root)
        .assert()
        .failure()
        .stderr(predicate::str::contains(
            "failed to determine the current Cargo project",
        ));

    log::info!("given: book is a Cargo project");
    Command::new("cargo")
        .arg("init")
        .args(["--name", "temp"])
        .env("PATH", &path)
        .current_dir(&root)
        .assert()
        .success();

    if find_code_extension().is_some()
        && may_skip("rust-analyzer code extension is already installed")
    {
        log::info!("when: book has item links");
        std::fs::File::options()
            .append(true)
            .open(root.path().join("src/chapter_1.md"))?
            .pipe(|mut file| file.write_all("\n[std::thread]\n".as_bytes()))?;

        log::info!("then: book builds without errors");
        Command::new("mdbook")
            .arg("build")
            .env("PATH", &path)
            .current_dir(&root)
            .assert()
            .success();
    } else if Command::new("rust-analyzer")
        .arg("--version")
        .assert()
        .try_success()
        .is_ok()
        && may_skip("rust-analyzer is already available")
    {
        log::info!("skip testing mdbook build without rust-analyzer")
    } else {
        log::info!("when: rust-analyzer is not configured");

        log::info!("when: book has no item links");

        log::info!("then: book builds without errors");
        Command::new("mdbook")
            .arg("build")
            .env("PATH", &path)
            .current_dir(&root)
            .assert()
            .success();

        log::info!("when: book has item links");
        std::fs::File::options()
            .append(true)
            .open(root.path().join("src/chapter_1.md"))?
            .pipe(|mut file| file.write_all("\n[std]\n".as_bytes()))?;

        log::info!("then: preprocessor fails");
        Command::new("mdbook")
            .arg("build")
            .env("PATH", &path)
            .current_dir(&root)
            .assert()
            .failure()
            .stderr(
                predicate::str::contains("failed to spawn rust-analyzer")
                    // https://github.com/rust-lang/rustup/issues/3846
                    // rustup shims rust-analyzer when it's not installed
                    .or(predicate::str::contains("Unknown binary 'rust-analyzer")),
                // ^ doesn't have a closing `'` because on windows it says 'rust-analyzer.exe'
            );

        log::info!("when: code extension is installed");

        let extension_dir = tempfile::Builder::new()
            .prefix(".vscode")
            .suffix("")
            .rand_bytes(0)
            .tempdir_in(dirs::home_dir().context("failed to get home dir")?)?;

        let ra_executable = extension_dir
            .path()
            .join("extensions/rust-lang.rust-analyzer-lorem-ipsum")
            .join("server/rust-analyzer");

        Command::new("cargo")
            .args(["run", "--package", "util-rust-analyzer", "--"])
            .arg("--ra-path")
            .arg(ra_executable)
            .arg("download")
            .unwrap()
            .assert()
            .success();

        log::info!("then: book builds without errors");
        Command::new("mdbook")
            .arg("build")
            .env("PATH", &path)
            .current_dir(&root)
            .assert()
            .success();
    }

    Ok(())
}
