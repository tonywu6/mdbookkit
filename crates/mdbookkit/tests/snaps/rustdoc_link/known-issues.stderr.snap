---
source: crates/mdbookkit/tests/rustdoc_link.rs
expression: report
---
  info: successfully resolved all links
    ╭─[known-issues:45:3]
 44 │ 
 45 │ - [~~`panic!`~~], and many other `std` macros
    ·   ───────┬──────
    ·          ╰── https://doc.rust-lang.org/stable/std/macros/macro.panic.html
 46 │   - The correct link is
 47 │     [https://doc.rust-lang.org/stable/std~~/macros~~/macro.panic.html][panic]
 48 │ - [~~`serde_json::json!`~~]
    ·   ────────────┬────────────
    ·               ╰── https://docs.rs/serde_json/1.0.140/serde_json/macros/macro.json.html
 49 │   - The correct link is
 50 │     [https://docs.rs/serde_json/1.0.140/serde_json~~/macros~~/macro.json.html][serde_json::json]
 51 │ 
 52 │ Attribute macros generate links that use `macro.<macro_name>.html`, but rustdoc actually
 53 │ generates `attr.<macro_name>.html`. For example:
 54 │ 
 55 │ - [~~`tokio::main!`~~]
    ·   ──────────┬─────────
    ·             ╰── https://docs.rs/tokio-macros/2.5.0/tokio_macros/macro.main.html
 56 │   - The correct link is
 57 │     [https://docs.rs/tokio-macros/2.5.0/tokio_macros/~~macro~~attr.main.html][tokio::main]
 58 │ 
 59 │ ### Trait items
 60 │ 
 61 │ Rust allows methods to have the same name if they are from different traits, and types
 62 │ can implement the same trait multiple times if the trait is generic. All such methods
 63 │ will appear on the same page for the type.
 64 │ 
 65 │ rustdoc will number the generated URL fragments so that they remain unique within the
 66 │ HTML document. rust-analyzer does not yet have the ability to do so.
 67 │ 
 68 │ For example, these are the same links:
 69 │ 
 70 │ - [`<std::net::IpAddr as From<std::net::Ipv4Addr>>::from`]
    ·   ────────────────────────────┬───────────────────────────
    ·                               ╰── https://doc.rust-lang.org/stable/core/net/ip_addr/enum.IpAddr.html#method.from
 71 │ - [`<std::net::IpAddr as From<std::net::Ipv6Addr>>::from`]
    ·   ────────────────────────────┬───────────────────────────
    ·                               ╰── https://doc.rust-lang.org/stable/core/net/ip_addr/enum.IpAddr.html#method.from
 72 │ 
    ╰────
